File: tests/programmers/HeapTest.java
package programmers;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;
import programmers.heap.DiskController;
import programmers.heap.MoreSpicy;

import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.params.provider.Arguments.arguments;

class heapTest {
    @Test
    public void moreSpicyTest() {
        int result = new MoreSpicy().solution(new int[]{1, 2, 3, 9, 10, 12}, 7);
        assertEquals(2, result);
    }

    private static Stream<Arguments> diskControllerArgs() {
        return Stream.of(
                arguments(9, new int[][]{{0, 3}, {1, 9}, {2, 6}}),
                arguments(9, new int[][]{{1, 10}, {3, 3}, {10, 3}}),
                arguments(15, new int[][]{{0, 10}, {4, 10}, {5, 11}, {15, 2}}),
                arguments(10, new int[][]{{0, 10}}),
                arguments(9, new int[][]{{0, 3}, {1, 9}, {2, 6}, {4, 3}}),
                arguments(3, new int[][]{{0, 1}, {1, 2}, {500, 6}}),
                arguments(6, new int[][]{{0, 3}, {1, 9}, {500, 6}}),
                arguments(1, new int[][]{{0, 1}, {0, 1}, {1, 0}}),
                arguments(3, new int[][]{{0, 3}, {4, 3}, {8, 3}}),
                arguments(3, new int[][]{{0, 3}, {4, 3}, {10, 3}}),
                arguments(3, new int[][]{{0, 5}, {6, 2}, {6, 1}}),
                arguments(3, new int[][]{{0, 5}, {6, 1}, {6, 2}}),
                arguments(5, new int[][]{{0, 5}, {2, 2}, {5, 3}}),
                arguments(5, new int[][]{{0, 5}, {2, 2}, {4, 2}}),
                arguments(4, new int[][]{{0, 3}, {0, 1}, {4, 7}}),
                arguments(3, new int[][]{{0, 2}, {3, 6}, {3, 1}}),
                arguments(6, new int[][]{{0, 5}, {1, 2}, {5, 5}}),
                arguments(13, new int[][]{{0, 9}, {0, 4}, {0, 5}, {0, 7}, {0, 3}}),
                arguments(72, new int[][]{{24, 10}, {28, 39}, {43, 20}, {37, 5}, {47, 22}, {20, 47}, {15, 2}, {15, 34}, {35, 43}, {26, 1}}),
                arguments(72, new int[][]{{24, 10}, {28, 39}, {43, 20}, {37, 5}, {47, 22}, {20, 47}, {15, 34}, {15, 2}, {35, 43}, {26, 1}}),
                arguments(13, new int[][]{{1, 9}, {1, 4}, {1, 5}, {1, 7}, {1, 3}})
        );
    }

    @ParameterizedTest
    @MethodSource("diskControllerArgs")
    public void diskControllerTest(int result, int[][] jobs) {
        assertEquals(result, new DiskController().solution(jobs));
    }
}

File: src/test/kotlin/study/programmers/HeapTest.kt
package study.programmers

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import study.programmers.heap.DiskController
import study.programmers.heap.MoreSpicy
import study.programmers.heap.solveDoublePriorityQueue
import java.util.stream.Stream

class HeapTest {
    @Test
    fun moreSpicyTest() {
        val result = MoreSpicy().solution(listOf(1, 2, 3, 9, 10, 12).toIntArray(), 7)
        assertEquals(2, result)
    }

    companion object {
        @JvmStatic
        fun diskControllerArgs(): Stream<Arguments> = Stream.of(
            Arguments.of(9, arrayOf(intArrayOf(0, 3), intArrayOf(1, 9), intArrayOf(2, 6))),
            Arguments.of(9, arrayOf(intArrayOf(1, 10), intArrayOf(3, 3), intArrayOf(10, 3))),
            Arguments.of(15, arrayOf(intArrayOf(0, 10), intArrayOf(4, 10), intArrayOf(5, 11), intArrayOf(15, 2))),
            Arguments.of(10, arrayOf(intArrayOf(0, 10))),
            Arguments.of(9, arrayOf(intArrayOf(0, 3), intArrayOf(1, 9), intArrayOf(2, 6), intArrayOf(4, 3))),
            Arguments.of(3, arrayOf(intArrayOf(0, 1), intArrayOf(1, 2), intArrayOf(500, 6))),
            Arguments.of(6, arrayOf(intArrayOf(0, 3), intArrayOf(1, 9), intArrayOf(500, 6))),
            Arguments.of(1, arrayOf(intArrayOf(0, 1), intArrayOf(0, 1), intArrayOf(1, 0))),
            Arguments.of(3, arrayOf(intArrayOf(0, 3), intArrayOf(4, 3), intArrayOf(8, 3))),
            Arguments.of(3, arrayOf(intArrayOf(0, 3), intArrayOf(4, 3), intArrayOf(10, 3))),
            Arguments.of(3, arrayOf(intArrayOf(0, 5), intArrayOf(6, 2), intArrayOf(6, 1))),
            Arguments.of(3, arrayOf(intArrayOf(0, 5), intArrayOf(6, 1), intArrayOf(6, 2))),
            Arguments.of(5, arrayOf(intArrayOf(0, 5), intArrayOf(2, 2), intArrayOf(5, 3))),
            Arguments.of(5, arrayOf(intArrayOf(0, 5), intArrayOf(2, 2), intArrayOf(4, 2))),
            Arguments.of(4, arrayOf(intArrayOf(0, 3), intArrayOf(0, 1), intArrayOf(4, 7))),
            Arguments.of(3, arrayOf(intArrayOf(0, 2), intArrayOf(3, 6), intArrayOf(3, 1))),
            Arguments.of(6, arrayOf(intArrayOf(0, 5), intArrayOf(1, 2), intArrayOf(5, 5))),
            Arguments.of(
                13,
                arrayOf(intArrayOf(0, 9), intArrayOf(0, 4), intArrayOf(0, 5), intArrayOf(0, 7), intArrayOf(0, 3))
            ),
            Arguments.of(
                72,
                arrayOf(
                    intArrayOf(24, 10),
                    intArrayOf(28, 39),
                    intArrayOf(43, 20),
                    intArrayOf(37, 5),
                    intArrayOf(47, 22),
                    intArrayOf(20, 47),
                    intArrayOf(15, 2),
                    intArrayOf(15, 34),
                    intArrayOf(35, 43),
                    intArrayOf(26, 1)
                )
            ),
            Arguments.of(
                72,
                arrayOf(
                    intArrayOf(24, 10),
                    intArrayOf(28, 39),
                    intArrayOf(43, 20),
                    intArrayOf(37, 5),
                    intArrayOf(47, 22),
                    intArrayOf(20, 47),
                    intArrayOf(15, 34),
                    intArrayOf(15, 2),
                    intArrayOf(35, 43),
                    intArrayOf(26, 1)
                )
            ),
            Arguments.of(
                13, arrayOf(intArrayOf(1, 9), intArrayOf(1, 4), intArrayOf(1, 5), intArrayOf(1, 7), intArrayOf(1, 3))
            )
        )

        @JvmStatic
        fun doublePriorityQueueArgs(): Stream<Arguments> = Stream.of(
            Arguments.of(arrayOf("I 16", "D 1"), intArrayOf(0, 0)),
            Arguments.of(arrayOf("I 7", "I 5", "I -5", "D -1"), intArrayOf(7, 5))
        )
    }

    @ParameterizedTest
    @MethodSource("diskControllerArgs")
    fun diskControllerTest(result: Int, jobs: Array<IntArray>) {
        assertEquals(result, DiskController().solution(jobs))
    }

    @ParameterizedTest
    @MethodSource("doublePriorityQueueArgs")
    fun testDoublePriorityQueue(operations: Array<String>, expected: IntArray) {
        val result = solveDoublePriorityQueue(operations)
        assertEquals(expected[0], result[0])
        assertEquals(expected[1], result[1])
    }
}
